src/utils.rs

```diff
@@ -167,3 +168,69 @@ impl<'a, T, const N: usize> IntoOption<&'a [T]> for &'a [T; N] {
         Some(self)
     }
 }
+
+pub trait ScalarOrArray<'a> {
+    type Array: AsRef<Array> + 'a;
+
+    fn into_owned_or_ref_array(self) -> Self::Array;
+}
+
+impl<'a> ScalarOrArray<'a> for Array {
+    type Array = Array;
+
+    fn into_owned_or_ref_array(self) -> Array {
+        self
+    }
+}
+
+impl<'a> ScalarOrArray<'a> for &'a Array {
+    type Array = &'a Array;
+
+    // TODO: clippy would complain about `as_array`. Is there a better name?
+    fn into_owned_or_ref_array(self) -> &'a Array {
+        self
+    }
+}
+
+impl ScalarOrArray<'static> for bool {
+    type Array = Array;
+
+    fn into_owned_or_ref_array(self) -> Array {
+        Array::from_bool(self)
+    }
+}
+
+impl ScalarOrArray<'static> for i32 {
+    type Array = Array;
+
+    fn into_owned_or_ref_array(self) -> Array {
+        Array::from_int(self)
+    }
+}
+
+impl ScalarOrArray<'static> for f32 {
+    type Array = Array;
+
+    fn into_owned_or_ref_array(self) -> Array {
+        Array::from_float(self)
+    }
+}
```

Comment:
can't you do something like what i've done in `random`? 
```
impl<T: ArrayElement> From<T> for ScalarOrArray {
```

Reply:
There's a tricky case. A scalar array is treated differently from a singleton matrix (one-element array). One example (but not exhaustive) that I'm aware of right now is `inner` product can broadcast if one arg is a scalar but fails if it's a singleton matrix.

For the scalars, mlx-c only provides bindings for `i32`, `bool`, and `f32`. So it's not as simple as a plain `Into`

Reply:
For the case with `inner`, the same behavior is observed in numpy, so I don't think that is a bug